allennlp/data/transforms.py

```diff
@@ -0,0 +1,531 @@
+        # HACK, just so we can use the existing sort_by_padding,
+        # only works if instances are indexed already.
+        self.vocab = None
+
+    def transform(self, dataset: Iterable[Instance]) -> Iterable[Batched]:
+
+        instances = list(dataset)
+        if not all([i.indexed for i in instances]):
+            raise ValueError("Index() must occur before SortByPadding()")
+
+        # TODO(Mark): Move this to somewhere where it is importable at the top level
+        from allennlp.data.iterators.bucket_iterator import (
+            sort_by_padding as allennlp_sort_by_padding,
+        )
+
+        instances = allennlp_sort_by_padding(
+            instances, self.sorting_keys, self.vocab, self.padding_noise
+        )
+
+        yield instances
+
+
+@Transform.register("epoch_tracker")
+class EpochTracker(Transform[Instance]):
+    """
+    Adds a allennlp Field to each Instance which specifies how many
+    times the full dataset has been iterated over.
+    """
+
+    def __init__(self):
+        self.epoch = 0
+
+    def transform(self, dataset: Iterable[Instance]) -> Iterable[Instance]:
+
+        for instance in dataset:
+            instance.fields["epoch_num"] = MetadataField(self.epoch)
+            yield instance
+        self.epoch += 1
+
+
+@Transform.register("skip_smaller_than")
+class SkipSmallerThan(Transform[Batched]):
+    """
+    Skip batches that are smaller than a specified size.
+    Useful if you don't want the uneven tail of a dataset
+    to be returned as a smaller batch.
+
+    Parameters
+    ----------
+    min_size : int
+        The minimum size of a batch. Only batches of size greater
+        than or equal to this will be returned.
+    """
+
+    def __init__(self, min_size: int):
+        self.min_size = min_size
+
+    def transform(self, dataset: Iterable[Instance]) -> Iterable[Batched]:
+
+        raise NotImplementedError(
+            "SkipSmallerThan cannot be called on Instances. Please add Batch to your pipeline first."
+        )
+
+    def transform_batch(self, batches: Iterable[Batched]) -> Iterable[Batched]:
+        for batch in batches:
+            batch = list(batch)
+
+            if len(batch) >= self.min_size:
+                yield batch
+
+
+@Transform.register("stop_after")
+class StopAfter(Transform[Instance]):
+    """
+    Stop an epoch after a fixed number of individual instances.
+    NOTE: The TransformIterator will ensure that the next epoch
+    starts from where we stopped.
+
+    Parameters
+    ----------
+    max : int
+        The number of instances to yield in a given epoch.
+    """
+
+    def __init__(self, max: int):
+        self.max = max
+
+    def transform(self, dataset: Iterable[Instance]) -> Iterable[Instance]:
+        i = 0
+        for instance in dataset:
+            yield instance
+            i += 1
+            if i >= self.max:
+                break
+
+    def transform_batch(self, batches: Iterable[Batched]) -> Iterable[Batched]:
```

Comment:
I think I'm missing something here. Why does `StopAfter` operate differently in the `transform` and `transform_batch` functions?

Reply:
The default implementation of the `transform_batch` is to apply `transform` to each batch passed to `transform_batch` and `yield from`, resulting in a flattened stream (which might still be batched, because some transforms map from `Instances -> Batched`). So transform_batch in this case is actually implementing `StopAfter` at the batch level, which is what would be expected. 

I agree it's a bit odd that this is the default implementation (rather than assuming that `transform_batch` would apply `transform` and just `yield`) - see [this comment](https://github.com/allenai/allennlp/pull/3386/files#diff-41d08074b6dcde4374683117cb26a2cbR46). I did it this way around just because more of the `Transforms` use this than don't, which resulted in less of them needing to override the method. I can switch it around if you want.

Reply:
It's possible that you might want to base classes here, one for instance -> instance transforms, and one for instance -> batch transforms, that have different default implementations of `transform_batch`.

Note that this is not in conflict with my saying that we should not have separate `Transform` and `InstanceTransform` classes.  That complaint is about the *type system* and how it interacts with `FromParams` and such - it's really annoying and adds a bunch of cognitive load to see (and use) `List[InstanceTransform]` in my type annotations for the data iterators, instead of just `List[Transform]` (i.e., I have to think about what the difference is, why there's a difference, and which one I need to use).   I think using `List[InstanceTransform]` in the data iterators even precludes the advanced usage you were talking about, like passing in strings and transforming them to lists of instances.